// electron-rebuild/lib/rebuild.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rebuildNativeModules = undefined;

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _bluebird = require('bluebird');

var _spawnRx = require('spawn-rx');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _fsPromise = require('fs-promise');

var _fsPromise2 = _interopRequireDefault(_fsPromise);

var _nodeAbi = require('node-abi');

var _nodeAbi2 = _interopRequireDefault(_nodeAbi);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _readPackageJson = require('./read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var d = (0, _debug2.default)('electron-rebuild');

var defaultMode = process.platform === 'win32' ? 'sequential' : 'parallel';

var _rebuild = function () {
  var _ref = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee5(lifecycle, buildPath, electronVersion) {
    var arch = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : process.arch;
    var extraModules = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : [];
    var forceRebuild = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : false;
    var headerURL = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : 'https://atom.io/download/electron';
    var types = arguments.length > 7 && arguments[7] !== undefined ? arguments[7] : ['prod', 'optional'];
    var mode = arguments.length > 8 && arguments[8] !== undefined ? arguments[8] : defaultMode;

    var prodDeps, rebuilds, ABI, rebuildModuleAt, rebuildAllModulesIn, findModule, markChildrenAsProdDeps, rootPackageJSON, markWaiters, depKeys, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, rebuildFn;

    return _regenerator2.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            if (_path2.default.isAbsolute(buildPath)) {
              _context5.next = 2;
              break;
            }

            throw new Error('Expected buildPath to be an absolute path');

          case 2:
            d('rebuilding with args:', buildPath, electronVersion, arch, extraModules, forceRebuild, headerURL, types);
            prodDeps = {};
            rebuilds = [];
            ABI = _nodeAbi2.default.getAbi(electronVersion, 'electron');


            (extraModules || []).forEach(function (moduleName) {
              if (!moduleName) return;
              prodDeps[moduleName] = true;
            });

            lifecycle.emit('start');

            rebuildModuleAt = function () {
              var _ref2 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee2(modulePath) {
                var _ret;

                return _regenerator2.default.wrap(function _callee2$(_context2) {
                  while (1) {
                    switch (_context2.prev = _context2.next) {
                      case 0:
                        _context2.next = 2;
                        return _fsPromise2.default.exists(_path2.default.resolve(modulePath, 'binding.gyp'));

                      case 2:
                        if (!_context2.sent) {
                          _context2.next = 7;
                          break;
                        }

                        return _context2.delegateYield(_regenerator2.default.mark(function _callee() {
                          var metaPath, meta, rebuildArgs, modulePackageJSON, moduleName, abiPath, nodePath;
                          return _regenerator2.default.wrap(function _callee$(_context) {
                            while (1) {
                              switch (_context.prev = _context.next) {
                                case 0:
                                  metaPath = _path2.default.resolve(modulePath, 'build', 'Release', '.forge-meta');

                                  lifecycle.emit('module-found', _path2.default.basename(modulePath));
                                  _context.t0 = !forceRebuild;

                                  if (!_context.t0) {
                                    _context.next = 7;
                                    break;
                                  }

                                  _context.next = 6;
                                  return _fsPromise2.default.exists(metaPath);

                                case 6:
                                  _context.t0 = _context.sent;

                                case 7:
                                  if (!_context.t0) {
                                    _context.next = 16;
                                    break;
                                  }

                                  _context.next = 10;
                                  return _fsPromise2.default.readFile(metaPath, 'utf8');

                                case 10:
                                  meta = _context.sent;

                                  if (!(meta === arch)) {
                                    _context.next = 16;
                                    break;
                                  }

                                  d('skipping: ' + _path2.default.basename(modulePath) + ' as it is already built');
                                  lifecycle.emit('module-done');
                                  lifecycle.emit('module-skip');
                                  return _context.abrupt('return', {
                                    v: void 0
                                  });

                                case 16:
                                  d('rebuilding:', _path2.default.basename(modulePath));
                                  rebuildArgs = ['rebuild', '--target=' + electronVersion, '--arch=' + arch, '--dist-url=' + headerURL, '--build-from-source'];
                                  _context.next = 20;
                                  return (0, _readPackageJson2.default)(modulePath);

                                case 20:
                                  modulePackageJSON = _context.sent;

                                  (0, _keys2.default)(modulePackageJSON.binary || {}).forEach(function (binaryKey) {
                                    var value = modulePackageJSON.binary[binaryKey];
                                    if (binaryKey === 'module_path') {
                                      value = _path2.default.resolve(modulePath, value);
                                    }
                                    rebuildArgs.push('--' + binaryKey + '=' + value);
                                  });

                                  _context.next = 24;
                                  return (0, _spawnRx.spawnPromise)(_path2.default.resolve(__dirname, '../node_modules/.bin/node-gyp' + (process.platform === 'win32' ? '.cmd' : '')), rebuildArgs, {
                                    cwd: modulePath,
                                    env: (0, _assign2.default)({}, process.env, {
                                      HOME: _path2.default.resolve(_os2.default.homedir(), '.electron-gyp'),
                                      USERPROFILE: _path2.default.resolve(_os2.default.homedir(), '.electron-gyp'),
                                      npm_config_disturl: 'https://atom.io/download/electron',
                                      npm_config_runtime: 'electron',
                                      npm_config_arch: arch,
                                      npm_config_target_arch: arch,
                                      npm_config_build_from_source: true
                                    })
                                  });

                                case 24:

                                  d('built:', _path2.default.basename(modulePath));
                                  _context.next = 27;
                                  return _fsPromise2.default.mkdirs(_path2.default.dirname(metaPath));

                                case 27:
                                  _context.next = 29;
                                  return _fsPromise2.default.writeFile(metaPath, arch);

                                case 29:

                                  d('copying to prebuilt place:', _path2.default.basename(modulePath));
                                  moduleName = _path2.default.basename(modulePath);
                                  abiPath = _path2.default.resolve(modulePath, 'bin/' + process.platform + '-' + arch + '-' + ABI);
                                  nodePath = _path2.default.resolve(modulePath, 'build/Release/' + moduleName + '.node');
                                  _context.next = 35;
                                  return _fsPromise2.default.exists(nodePath);

                                case 35:
                                  if (!_context.sent) {
                                    _context.next = 40;
                                    break;
                                  }

                                  _context.next = 38;
                                  return _fsPromise2.default.mkdirs(abiPath);

                                case 38:
                                  _context.next = 40;
                                  return _fsPromise2.default.copy(nodePath, _path2.default.resolve(abiPath, moduleName + '.node'));

                                case 40:
                                  lifecycle.emit('module-done');

                                case 41:
                                case 'end':
                                  return _context.stop();
                              }
                            }
                          }, _callee, undefined);
                        })(), 't0', 4);

                      case 4:
                        _ret = _context2.t0;

                        if (!((typeof _ret === 'undefined' ? 'undefined' : (0, _typeof3.default)(_ret)) === "object")) {
                          _context2.next = 7;
                          break;
                        }

                        return _context2.abrupt('return', _ret.v);

                      case 7:
                      case 'end':
                        return _context2.stop();
                    }
                  }
                }, _callee2, undefined);
              }));

              return function rebuildModuleAt(_x10) {
                return _ref2.apply(this, arguments);
              };
            }();

            rebuildAllModulesIn = function rebuildAllModulesIn(nodeModulesPath) {
              var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

              d('scanning:', nodeModulesPath);
              var _iteratorNormalCompletion = true;
              var _didIteratorError = false;
              var _iteratorError = undefined;

              try {
                var _loop = function _loop() {
                  var modulePath = _step.value;

                  if (prodDeps['' + prefix + modulePath]) {
                    rebuilds.push(function () {
                      return rebuildModuleAt(_path2.default.resolve(nodeModulesPath, modulePath));
                    });
                  }
                  if (modulePath.startsWith('@')) {
                    rebuildAllModulesIn(_path2.default.resolve(nodeModulesPath, modulePath), modulePath + '/');
                  }
                  if (_fsPromise2.default.existsSync(_path2.default.resolve(nodeModulesPath, modulePath, 'node_modules'))) {
                    rebuildAllModulesIn(_path2.default.resolve(nodeModulesPath, modulePath, 'node_modules'));
                  }
                };

                for (var _iterator = (0, _getIterator3.default)(_fsPromise2.default.readdirSync(nodeModulesPath)), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                  _loop();
                }
              } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
              } finally {
                try {
                  if (!_iteratorNormalCompletion && _iterator.return) {
                    _iterator.return();
                  }
                } finally {
                  if (_didIteratorError) {
                    throw _iteratorError;
                  }
                }
              }
            };

            findModule = function () {
              var _ref3 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee3(moduleName, fromDir, foundFn) {
                var targetDir, foundFns, testPath;
                return _regenerator2.default.wrap(function _callee3$(_context3) {
                  while (1) {
                    switch (_context3.prev = _context3.next) {
                      case 0:
                        targetDir = fromDir;
                        foundFns = [];

                      case 2:
                        if (!(targetDir !== buildPath)) {
                          _context3.next = 11;
                          break;
                        }

                        testPath = _path2.default.resolve(targetDir, 'node_modules', moduleName);
                        _context3.next = 6;
                        return _fsPromise2.default.exists(testPath);

                      case 6:
                        if (!_context3.sent) {
                          _context3.next = 8;
                          break;
                        }

                        foundFns.push(foundFn(testPath));

                      case 8:
                        targetDir = _path2.default.dirname(targetDir);
                        _context3.next = 2;
                        break;

                      case 11:
                        _context3.next = 13;
                        return _promise2.default.all(foundFns);

                      case 13:
                      case 'end':
                        return _context3.stop();
                    }
                  }
                }, _callee3, undefined);
              }));

              return function findModule(_x12, _x13, _x14) {
                return _ref3.apply(this, arguments);
              };
            }();

            markChildrenAsProdDeps = function () {
              var _ref4 = (0, _bluebird.coroutine)(_regenerator2.default.mark(function _callee4(modulePath) {
                var childPackageJSON, moduleWait;
                return _regenerator2.default.wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        _context4.next = 2;
                        return _fsPromise2.default.exists(modulePath);

                      case 2:
                        if (_context4.sent) {
                          _context4.next = 4;
                          break;
                        }

                        return _context4.abrupt('return');

                      case 4:
                        _context4.next = 6;
                        return (0, _readPackageJson2.default)(modulePath);

                      case 6:
                        childPackageJSON = _context4.sent;
                        moduleWait = [];


                        (0, _keys2.default)(childPackageJSON.dependencies || {}).concat((0, _keys2.default)(childPackageJSON.optionalDependencies || {})).forEach(function (key) {
                          if (prodDeps[key]) return;
                          prodDeps[key] = true;
                          moduleWait.push(findModule(key, modulePath, markChildrenAsProdDeps));
                        });
                        _context4.next = 11;
                        return _promise2.default.all(moduleWait);

                      case 11:
                      case 'end':
                        return _context4.stop();
                    }
                  }
                }, _callee4, undefined);
              }));

              return function markChildrenAsProdDeps(_x15) {
                return _ref4.apply(this, arguments);
              };
            }();

            _context5.next = 14;
            return (0, _readPackageJson2.default)(buildPath);

          case 14:
            rootPackageJSON = _context5.sent;
            markWaiters = [];
            depKeys = [];

            if (types.indexOf('prod') !== -1) {
              depKeys.push.apply(depKeys, (0, _toConsumableArray3.default)((0, _keys2.default)(rootPackageJSON.dependencies || {})));
            }
            if (types.indexOf('optional') !== -1) {
              depKeys.push.apply(depKeys, (0, _toConsumableArray3.default)((0, _keys2.default)(rootPackageJSON.optionalDependencies || {})));
            }
            if (types.indexOf('dev') !== -1) {
              depKeys.push.apply(depKeys, (0, _toConsumableArray3.default)((0, _keys2.default)(rootPackageJSON.devDependencies || {})));
            }
            depKeys.forEach(function (key) {
              prodDeps[key] = true;
              markWaiters.push(markChildrenAsProdDeps(_path2.default.resolve(buildPath, 'node_modules', key)));
            });

            _context5.next = 23;
            return _promise2.default.all(markWaiters);

          case 23:

            d('identified prod deps:', prodDeps);

            rebuildAllModulesIn(_path2.default.resolve(buildPath, 'node_modules'));

            if (!(mode !== 'sequential')) {
              _context5.next = 30;
              break;
            }

            _context5.next = 28;
            return _promise2.default.all(rebuilds.map(function (fn) {
              return fn();
            }));

          case 28:
            _context5.next = 56;
            break;

          case 30:
            _iteratorNormalCompletion2 = true;
            _didIteratorError2 = false;
            _iteratorError2 = undefined;
            _context5.prev = 33;
            _iterator2 = (0, _getIterator3.default)(rebuilds);

          case 35:
            if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
              _context5.next = 42;
              break;
            }

            rebuildFn = _step2.value;
            _context5.next = 39;
            return rebuildFn();

          case 39:
            _iteratorNormalCompletion2 = true;
            _context5.next = 35;
            break;

          case 42:
            _context5.next = 48;
            break;

          case 44:
            _context5.prev = 44;
            _context5.t0 = _context5['catch'](33);
            _didIteratorError2 = true;
            _iteratorError2 = _context5.t0;

          case 48:
            _context5.prev = 48;
            _context5.prev = 49;

            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }

          case 51:
            _context5.prev = 51;

            if (!_didIteratorError2) {
              _context5.next = 54;
              break;
            }

            throw _iteratorError2;

          case 54:
            return _context5.finish(51);

          case 55:
            return _context5.finish(48);

          case 56:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee5, undefined, [[33, 44, 48, 56], [49,, 51, 55]]);
  }));

  return function _rebuild(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

var rebuild = function rebuild() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  var lifecycle = new _events2.default();
  d('rebuilding with args:', args);
  var rebuilder = _rebuild.apply(undefined, [lifecycle].concat(args));
  rebuilder.lifecycle = lifecycle;
  return rebuilder;
};

var rebuildNativeModules = exports.rebuildNativeModules = function rebuildNativeModules(electronVersion, modulePath) {
  var whichModule = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
  var headersDir = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
  var arch = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : process.arch;
  var command = arguments[5];
  var ignoreDevDeps = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : false;
  var ignoreOptDeps = arguments.length > 7 && arguments[7] !== undefined ? arguments[7] : false;
  var verbose = arguments.length > 8 && arguments[8] !== undefined ? arguments[8] : false;

  d('rebuilding in:', modulePath);
  console.warn('You are using the old API, please read the new docs and update to thew new API');
  return rebuild(modulePath, electronVersion, arch, whichModule.split(','));
};

exports.default = rebuild;
